cluster/manager.go

```diff
@@ -0,0 +1,229 @@
+package cluster
+
+import (
+	"encoding/json"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/memberlist"
+	"github.com/raintank/worldping-api/pkg/log"
+)
+
+type ClusterManager struct {
+	sync.RWMutex
+	Peers map[string]Node
+	node  Node
+	list  *memberlist.Memberlist
+}
+
+func (c *ClusterManager) SetList(list *memberlist.Memberlist) {
+	c.Lock()
+	c.list = list
+	c.Unlock()
+}
+
+func (c *ClusterManager) ThisNode() Node {
+	c.RLock()
+	defer c.RUnlock()
+	return c.node
+}
+
+func (c *ClusterManager) PeersList() []Node {
+	c.RLock()
+	list := make([]Node, len(c.Peers))
+	i := 0
+	for _, n := range c.Peers {
+		list[i] = n
+		i++
+	}
+	c.RUnlock()
+	return list
+}
+
+func (c *ClusterManager) NotifyJoin(node *memberlist.Node) {
+	c.Lock()
+	defer c.Unlock()
+	if len(node.Meta) == 0 {
+		return
+	}
+	log.Info("Node %s with address %s has joined the cluster\n", node.Name, node.Addr.String())
+	peer := Node{}
+	err := json.Unmarshal(node.Meta, &peer)
+	if err != nil {
+		panic(err)
+	}
+	peer.RemoteAddr = node.Addr.String()
+	if peer.Name == c.node.Name {
+		peer.local = true
+	}
+	c.Peers[node.Name] = peer
+}
+
+func (c *ClusterManager) NotifyLeave(node *memberlist.Node) {
+	c.Lock()
+	defer c.Unlock()
+	log.Info("Node %s has left the cluster\n", node.Name)
+	delete(c.Peers, node.Name)
+}
+
+func (c *ClusterManager) NotifyUpdate(node *memberlist.Node) {
+	c.Lock()
+	defer c.Unlock()
+	if len(node.Meta) == 0 {
+		return
+	}
+	peer := Node{}
+	err := json.Unmarshal(node.Meta, &peer)
+	if err != nil {
+		panic(err)
```

Comment:
should we panic? seems like 1 MT instance sending faulty data can take down an entire cluster.
not sure how to properly handle though, ideally I think we simply wouldn't update our info, but we should probably also mark it as faulty.